<p style="text-align:center">
    <a href="https://skills.network/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01" target="_blank">
    <img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/assets/logos/SN_web_lightmode.png" width="200" alt="Skills Network Logo"  />
    </a>
</p>


<h1>Extracting and Visualizing Stock Data</h1>
<h2>Description</h2>


Extracting essential data from a dataset and displaying it is a necessary part of data science; therefore individuals can make correct decisions based on the data. In this assignment, you will extract some stock data, you will then display this data in a graph.


<h2>Table of Contents</h2>
<div class="alert alert-block alert-info" style="margin-top: 20px">
    <ul>
        <li>Define a Function that Makes a Graph</li>
        <li>Question 1: Use yfinance to Extract Stock Data</li>
        <li>Question 2: Use Webscraping to Extract Tesla Revenue Data</li>
        <li>Question 3: Use yfinance to Extract Stock Data</li>
        <li>Question 4: Use Webscraping to Extract GME Revenue Data</li>
        <li>Question 5: Plot Tesla Stock Graph</li>
        <li>Question 6: Plot GameStop Stock Graph</li>
    </ul>
<p>
    Estimated Time Needed: <strong>30 min</strong></p>
</div>

<hr>


In [ ]:
!pip install yfinance
#!pip install pandas
#!pip install requests
!pip install bs4
#!pip install plotly

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for bs4: filename=bs4-0.0.1-py3-none-any.whl size=1270 sha256=136a843c1bcaa4619b4cd8d66c62cefa1885f919bb9f0ad8293970b144fca38d
  Stored in directory: /root/.cache/pip/wheels/73/2b/cb/099980278a0c9a3e57ff1a89875ec07bfa0b6fcbebb9a8cad3
Successfully built bs4


In [ ]:
import yfinance as yf
import pandas as pd
import requests
from bs4 import BeautifulSoup
import plotly.graph_objects as go
from plotly.subplots import make_subplots

## Define Graphing Function


In this section, we define the function `make_graph`. You don't have to know how the function works, you should only care about the inputs. It takes a dataframe with stock data (dataframe must contain Date and Close columns), a dataframe with revenue data (dataframe must contain Date and Revenue columns), and the name of the stock.


In [ ]:
def make_graph(stock_data, revenue_data, stock):
    fig = make_subplots(rows=2, cols=1, shared_xaxes=True, subplot_titles=("Historical Share Price", "Historical Revenue"), vertical_spacing = .3)
    stock_data_specific = stock_data[stock_data.Date <= '2021--06-14']
    revenue_data_specific = revenue_data[revenue_data.Date <= '2021-04-30']
    fig.add_trace(go.Scatter(x=pd.to_datetime(stock_data_specific.Date, infer_datetime_format=True), y=stock_data_specific.Close.astype("float"), name="Share Price"), row=1, col=1)
    fig.add_trace(go.Scatter(x=pd.to_datetime(revenue_data_specific.Date, infer_datetime_format=True), y=revenue_data_specific.Revenue.astype("float"), name="Revenue"), row=2, col=1)
    fig.update_xaxes(title_text="Date", row=1, col=1)
    fig.update_xaxes(title_text="Date", row=2, col=1)
    fig.update_yaxes(title_text="Price ($US)", row=1, col=1)
    fig.update_yaxes(title_text="Revenue ($US Millions)", row=2, col=1)
    fig.update_layout(showlegend=False,
    height=900,
    title=stock,
    xaxis_rangeslider_visible=True)
    fig.show()

## Question 1: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is Tesla and its ticker symbol is `TSLA`.


In [ ]:
tesla = yf.Ticker('TSLA')

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `tesla_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [ ]:
tesla_data = tesla.history(period = 'max')

**Reset the index** using the `reset_index(inplace=True)` function on the tesla_data DataFrame and display the first five rows of the `tesla_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 1 to the results below.


In [ ]:
tesla_data = tesla.history(period = 'max')

## Question 2: Use Webscraping to Extract Tesla Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/revenue.htm Save the text of the response as a variable named `html_data`.


In [ ]:
tesla_data.reset_index(inplace=True)
tesla_data.head()

Date      Open      High       Low     Close  \
0 2010-06-29 00:00:00-04:00  1.266667  1.666667  1.169333  1.592667   
1 2010-06-30 00:00:00-04:00  1.719333  2.028000  1.553333  1.588667   
2 2010-07-01 00:00:00-04:00  1.666667  1.728000  1.351333  1.464000   
3 2010-07-02 00:00:00-04:00  1.533333  1.540000  1.247333  1.280000   
4 2010-07-06 00:00:00-04:00  1.333333  1.333333  1.055333  1.074000   

      Volume  Dividends  Stock Splits  
0  281494500        0.0           0.0  
1  257806500        0.0           0.0  
2  123282000        0.0           0.0  
3   77097000        0.0           0.0  
4  103003500        0.0           0.0

Parse the html data using `beautiful_soup`.


In [ ]:
url = "https://www.macrotrends.net/stocks/charts/TSLA/tesla/revenue"
html_data = requests.get(url).text
soup = BeautifulSoup(html_data, "html5lib")

Using `BeautifulSoup` or the `read_html` function extract the table with `Tesla Quarterly Revenue` and store it into a dataframe named `tesla_revenue`. The dataframe should have columns `Date` and `Revenue`.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [ ]:
tables = soup.find_all('table')
for index,table in enumerate(tables):
  if("Tesla Quarterly Revenue" in str(table)):
    table_index = index
print(table_index)

1


In [ ]:
tesla_revenue = pd.DataFrame(columns=['Date','Revenue'])
for row in tables[table_index].tbody.find_all('tr'):
  col = row.find_all('td')
  if(col != []):
    date = col[0].text
    revenue = col[1].text
    tesla_revenue = tesla_revenue.append({'Date':date, 'Revenue':revenue},ignore_index=True)
tesla_revenue

<ipython-input-9-a200ed229615>:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  tesla_revenue = tesla_revenue.append({'Date':date, 'Revenue':revenue},ignore_index=True)
<ipython-input-9-a200ed229615>:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  tesla_revenue = tesla_revenue.append({'Date':date, 'Revenue':revenue},ignore_index=True)
<ipython-input-9-a200ed229615>:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  tesla_revenue = tesla_revenue.append({'Date':date, 'Revenue':revenue},ignore_index=True)
<ipython-input-9-a200ed229615>:7: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  tesla_revenue = tesla_revenue.append({'Date':date, 'Reven

Date  Revenue
0   2022-12-31  $24,318
1   2022-09-30  $21,454
2   2022-06-30  $16,934
3   2022-03-31  $18,756
4   2021-12-31  $17,719
5   2021-09-30  $13,757
6   2021-06-30  $11,958
7   2021-03-31  $10,389
8   2020-12-31  $10,744
9   2020-09-30   $8,771
10  2020-06-30   $6,036
11  2020-03-31   $5,985
12  2019-12-31   $7,384
13  2019-09-30   $6,303
14  2019-06-30   $6,350
15  2019-03-31   $4,541
16  2018-12-31   $7,226
17  2018-09-30   $6,824
18  2018-06-30   $4,002
19  2018-03-31   $3,409
20  2017-12-31   $3,288
21  2017-09-30   $2,985
22  2017-06-30   $2,790
23  2017-03-31   $2,696
24  2016-12-31   $2,285
25  2016-09-30   $2,298
26  2016-06-30   $1,270
27  2016-03-31   $1,147
28  2015-12-31   $1,214
29  2015-09-30     $937
30  2015-06-30     $955
31  2015-03-31     $940
32  2014-12-31     $957
33  2014-09-30     $852
34  2014-06-30     $769
35  2014-03-31     $621
36  2013-12-31     $615
37  2013-09-30     $431
38  2013-06-30     $405
39  2013-03-31     $562
40  2012-12-31     $306
41  2012-09-30      $50
42  2012-06-30      $27
43  2012-03-31      $30
44  2011-12-31      $39
45  2011-09-30      $58
46  2011-06-30      $58
47  2011-03-31      $49
48  2010-12-31      $36
49  2010-09-30      $31
50  2010-06-30      $28
51  2010-03-31      $21
52  2009-12-31         
53  2009-09-30      $46
54  2009-06-30      $27

Execute the following line to remove the comma and dollar sign from the `Revenue` column. 


In [ ]:
tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"")
tesla_revenue

<ipython-input-10-6d89075fcbd5>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  tesla_revenue["Revenue"] = tesla_revenue['Revenue'].str.replace(',|\$',"")


Date Revenue
0   2022-12-31   24318
1   2022-09-30   21454
2   2022-06-30   16934
3   2022-03-31   18756
4   2021-12-31   17719
5   2021-09-30   13757
6   2021-06-30   11958
7   2021-03-31   10389
8   2020-12-31   10744
9   2020-09-30    8771
10  2020-06-30    6036
11  2020-03-31    5985
12  2019-12-31    7384
13  2019-09-30    6303
14  2019-06-30    6350
15  2019-03-31    4541
16  2018-12-31    7226
17  2018-09-30    6824
18  2018-06-30    4002
19  2018-03-31    3409
20  2017-12-31    3288
21  2017-09-30    2985
22  2017-06-30    2790
23  2017-03-31    2696
24  2016-12-31    2285
25  2016-09-30    2298
26  2016-06-30    1270
27  2016-03-31    1147
28  2015-12-31    1214
29  2015-09-30     937
30  2015-06-30     955
31  2015-03-31     940
32  2014-12-31     957
33  2014-09-30     852
34  2014-06-30     769
35  2014-03-31     621
36  2013-12-31     615
37  2013-09-30     431
38  2013-06-30     405
39  2013-03-31     562
40  2012-12-31     306
41  2012-09-30      50
42  2012-06-30      27
43  2012-03-31      30
44  2011-12-31      39
45  2011-09-30      58
46  2011-06-30      58
47  2011-03-31      49
48  2010-12-31      36
49  2010-09-30      31
50  2010-06-30      28
51  2010-03-31      21
52  2009-12-31        
53  2009-09-30      46
54  2009-06-30      27

Execute the following lines to remove an null or empty strings in the Revenue column.


In [ ]:
tesla_revenue.dropna(inplace=True)
tesla_revenue = tesla_revenue[tesla_revenue['Revenue'] != ""]
tesla_revenue

Date Revenue
0   2022-12-31   24318
1   2022-09-30   21454
2   2022-06-30   16934
3   2022-03-31   18756
4   2021-12-31   17719
5   2021-09-30   13757
6   2021-06-30   11958
7   2021-03-31   10389
8   2020-12-31   10744
9   2020-09-30    8771
10  2020-06-30    6036
11  2020-03-31    5985
12  2019-12-31    7384
13  2019-09-30    6303
14  2019-06-30    6350
15  2019-03-31    4541
16  2018-12-31    7226
17  2018-09-30    6824
18  2018-06-30    4002
19  2018-03-31    3409
20  2017-12-31    3288
21  2017-09-30    2985
22  2017-06-30    2790
23  2017-03-31    2696
24  2016-12-31    2285
25  2016-09-30    2298
26  2016-06-30    1270
27  2016-03-31    1147
28  2015-12-31    1214
29  2015-09-30     937
30  2015-06-30     955
31  2015-03-31     940
32  2014-12-31     957
33  2014-09-30     852
34  2014-06-30     769
35  2014-03-31     621
36  2013-12-31     615
37  2013-09-30     431
38  2013-06-30     405
39  2013-03-31     562
40  2012-12-31     306
41  2012-09-30      50
42  2012-06-30      27
43  2012-03-31      30
44  2011-12-31      39
45  2011-09-30      58
46  2011-06-30      58
47  2011-03-31      49
48  2010-12-31      36
49  2010-09-30      31
50  2010-06-30      28
51  2010-03-31      21
53  2009-09-30      46
54  2009-06-30      27

Display the last 5 row of the `tesla_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [ ]:
tesla_revenue.tail(5)

Date Revenue
49  2010-09-30      31
50  2010-06-30      28
51  2010-03-31      21
53  2009-09-30      46
54  2009-06-30      27

## Question 3: Use yfinance to Extract Stock Data


Using the `Ticker` function enter the ticker symbol of the stock we want to extract data on to create a ticker object. The stock is GameStop and its ticker symbol is `GME`.


In [ ]:
gamestop = yf.Ticker('GME')

Using the ticker object and the function `history` extract stock information and save it in a dataframe named `gme_data`. Set the `period` parameter to `max` so we get information for the maximum amount of time.


In [ ]:
gme_data = gamestop.history(period = 'max')

**Reset the index** using the `reset_index(inplace=True)` function on the gme_data DataFrame and display the first five rows of the `gme_data` dataframe using the `head` function. Take a screenshot of the results and code from the beginning of Question 3 to the results below.


In [ ]:
gme_data.reset_index(inplace=True)
gme_data.head()

Date      Open      High       Low     Close    Volume  \
0 2002-02-13 00:00:00-05:00  1.620128  1.693349  1.603295  1.691666  76216000   
1 2002-02-14 00:00:00-05:00  1.712708  1.716074  1.670626  1.683251  11021600   
2 2002-02-15 00:00:00-05:00  1.683250  1.687458  1.658002  1.674834   8389600   
3 2002-02-19 00:00:00-05:00  1.666418  1.666418  1.578047  1.607504   7410400   
4 2002-02-20 00:00:00-05:00  1.615920  1.662210  1.603296  1.662210   6892800   

   Dividends  Stock Splits  
0        0.0           0.0  
1        0.0           0.0  
2        0.0           0.0  
3        0.0           0.0  
4        0.0           0.0

## Question 4: Use Webscraping to Extract GME Revenue Data


Use the `requests` library to download the webpage https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-PY0220EN-SkillsNetwork/labs/project/stock.html. Save the text of the response as a variable named `html_data`.


In [ ]:
url = "https://www.macrotrends.net/stocks/charts/GME/gamestop/revenue"
html_data = requests.get(url).text

Parse the html data using `beautiful_soup`.


In [ ]:
soup = BeautifulSoup(html_data, "html5lib")

Using `BeautifulSoup` or the `read_html` function extract the table with `GameStop Quarterly Revenue` and store it into a dataframe named `gme_revenue`. The dataframe should have columns `Date` and `Revenue`. Make sure the comma and dollar sign is removed from the `Revenue` column using a method similar to what you did in Question 2.


<details><summary>Click here if you need help locating the table</summary>

```
    
Below is the code to isolate the table, you will now need to loop through the rows and columns like in the previous lab
    
soup.find_all("tbody")[1]
    
If you want to use the read_html function the table is located at index 1


```

</details>


In [ ]:
tables = soup.find_all('table')
for index,table in enumerate(tables):
  if("GameStop Quarterly Revenue" in str(table)):
    table_index = index

gme_revenue = pd.DataFrame(columns=['Date','Revenue'])
for row in tables[table_index].tbody.find_all('tr'):
  col = row.find_all('td')
  if(col != []):
    date = col[0].text
    revenue = col[1].text
    gme_revenue = gme_revenue.append({'Date':date, 'Revenue':revenue},ignore_index=True)
gme_revenue

<ipython-input-18-811423744643>:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  gme_revenue = gme_revenue.append({'Date':date, 'Revenue':revenue},ignore_index=True)
<ipython-input-18-811423744643>:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  gme_revenue = gme_revenue.append({'Date':date, 'Revenue':revenue},ignore_index=True)
<ipython-input-18-811423744643>:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  gme_revenue = gme_revenue.append({'Date':date, 'Revenue':revenue},ignore_index=True)
<ipython-input-18-811423744643>:12: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  gme_revenue = gme_revenue.append({'Date':date, 'Revenue':reve

Date Revenue
0   2023-01-31  $2,226
1   2022-10-31  $1,186
2   2022-07-31  $1,136
3   2022-04-30  $1,378
4   2022-01-31  $2,254
5   2021-10-31  $1,297
6   2021-07-31  $1,183
7   2021-04-30  $1,277
8   2021-01-31  $2,122
9   2020-10-31  $1,005
10  2020-07-31    $942
11  2020-04-30  $1,021
12  2020-01-31  $2,194
13  2019-10-31  $1,439
14  2019-07-31  $1,286
15  2019-04-30  $1,548
16  2019-01-31  $3,063
17  2018-10-31  $1,935
18  2018-07-31  $1,501
19  2018-04-30  $1,786
20  2018-01-31  $2,825
21  2017-10-31  $1,989
22  2017-07-31  $1,688
23  2017-04-30  $2,046
24  2017-01-31  $2,403
25  2016-10-31  $1,959
26  2016-07-31  $1,632
27  2016-04-30  $1,972
28  2016-01-31  $3,525
29  2015-10-31  $2,016
30  2015-07-31  $1,762
31  2015-04-30  $2,061
32  2015-01-31  $3,476
33  2014-10-31  $2,092
34  2014-07-31  $1,731
35  2014-04-30  $1,996
36  2014-01-31  $3,684
37  2013-10-31  $2,107
38  2013-07-31  $1,384
39  2013-04-30  $1,865
40  2013-01-31  $3,562
41  2012-10-31  $1,773
42  2012-07-31  $1,550
43  2012-04-30  $2,002
44  2012-01-31  $3,579
45  2011-10-31  $1,947
46  2011-07-31  $1,744
47  2011-04-30  $2,281
48  2011-01-31  $3,693
49  2010-10-31  $1,899
50  2010-07-31  $1,799
51  2010-04-30  $2,083
52  2010-01-31  $3,524
53  2009-10-31  $1,835
54  2009-07-31  $1,739
55  2009-04-30  $1,981
56  2009-01-31  $3,492

Display the last five rows of the `gme_revenue` dataframe using the `tail` function. Take a screenshot of the results.


In [ ]:
gme_revenue["Revenue"] = gme_revenue['Revenue'].str.replace(',|\$',"")
gme_revenue.dropna(inplace=True)
gme_revenue = gme_revenue[gme_revenue['Revenue'] != ""]
gme_revenue.tail(5)

<ipython-input-19-315b60cc1bb7>:1: FutureWarning: The default value of regex will change from True to False in a future version.
  gme_revenue["Revenue"] = gme_revenue['Revenue'].str.replace(',|\$',"")


Date Revenue
52  2010-01-31    3524
53  2009-10-31    1835
54  2009-07-31    1739
55  2009-04-30    1981
56  2009-01-31    3492

## Question 5: Plot Tesla Stock Graph


Use the `make_graph` function to graph the Tesla Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(tesla_data, tesla_revenue, 'Tesla')`. Note the graph will only show data upto June 2021.


In [ ]:
make_graph(tesla_data, tesla_revenue, 'Tesla')

## Question 6: Plot GameStop Stock Graph


Use the `make_graph` function to graph the GameStop Stock Data, also provide a title for the graph. The structure to call the `make_graph` function is `make_graph(gme_data, gme_revenue, 'GameStop')`. Note the graph will only show data upto June 2021.


In [ ]:
make_graph(gme_data, gme_revenue, 'GameStop')

<h2>About the Authors:</h2> 

<a href="https://www.linkedin.com/in/joseph-s-50398b136/?utm_medium=Exinfluencer&utm_source=Exinfluencer&utm_content=000026UJ&utm_term=10006555&utm_id=NA-SkillsNetwork-Channel-SkillsNetworkCoursesIBMDeveloperSkillsNetworkPY0220ENSkillsNetwork900-2022-01-01">Joseph Santarcangelo</a> has a PhD in Electrical Engineering, his research focused on using machine learning, signal processing, and computer vision to determine how videos impact human cognition. Joseph has been working for IBM since he completed his PhD.

Azim Hirjani


## Change Log

| Date (YYYY-MM-DD) | Version | Changed By    | Change Description        |
| ----------------- | ------- | ------------- | ------------------------- |
| 2022-02-28        | 1.2     | Lakshmi Holla | Changed the URL of GameStop |
| 2020-11-10        | 1.1     | Malika Singla | Deleted the Optional part |
| 2020-08-27        | 1.0     | Malika Singla | Added lab to GitLab       |

<hr>

## <h3 align="center"> © IBM Corporation 2020. All rights reserved. <h3/>

<p>
